### PR TITLE
Adding autocorrelation function to RTCC.

### PR DESCRIPTION
**Pull request by lothian**
## Description
This PR will add the autocorrelation function to RTCC to compute the population of the ground state as a function of time.

## Todos
  - [X] Add phase factor calculation to main RK f() function.
  - [x] Add function to compute autocorrelation function in each step
  - [x] Add test case comparing to Pedersen and Kvaal's results from [J. Chem. Phys. 150, 144106 (2019)](https://doi.org/10.1063/1.5085390).

## Other
  - [X] Removed the `energy` function from `rtcc.py` because it is unused and untested.
 
## Status
- [x] Ready to go

### CODE REVIEW

File: pycc/rt/rtcc.py [hunk 1]
```diff
@@ -126,42 +127,49 @@ def f(self, t, y):
...
+            return np.concatenate((t1, t2, l1, l2, phase), axis=None).astype('complex128')
 
 
     def extract_amps(self, y):
         """
         Parameters
         ----------
         y : NumPy array
-            flattened array of cluster amplitudes or residuals
+            flattened array of wave function phase and cluster amplitudes or residuals
```
Comment (bgpeyton) [inline pycc/rt/rtcc.py:167]: `amplitudes or residuals and phase as a vector (flattened array)` 1) for consistency with `collect_amps`, and 2) to imply `phase` is the last element in the vector

File: pycc/rt/rtcc.py [hunk 2]
```diff
@@ -175,14 +183,17 @@ def extract_amps(self, y):
...
             t2 = torch.reshape(y[len1:(len1+len2)], (no, no, nv, nv))
             l1 = torch.reshape(y[(len1+len2):(len1+len2+len1)], (no, nv))
-            l2 = torch.reshape(y[(len1+len2+len1):], (no, no, nv, nv))
+            l2 = torch.reshape(y[(len1+len2+len1):(2*len1+2*len2)], (no, no, nv, nv))
         else:
             t1 = np.reshape(y[:len1], (no, nv))
             t2 = np.reshape(y[len1:(len1+len2)], (no, no, nv, nv))
             l1 = np.reshape(y[(len1+len2):(len1+len2+len1)], (no, nv))
-            l2 = np.reshape(y[(len1+len2+len1):], (no, no, nv, nv))
+            l2 = np.reshape(y[(len1+len2+len1):(2*len1+2*len2)], (no, no, nv, nv))
```
Comment (bgpeyton) [inline pycc/rt/rtcc.py:191]: same as above

File: pycc/rt/rtcc.py [hunk 3]
```diff
@@ -175,14 +183,17 @@ def extract_amps(self, y):
...
 
-        return t1, t2, l1, l2
+        # Extract the phase
+        phase = y[y.size-1]
```
Comment (bgpeyton) [inline pycc/rt/rtcc.py:194]: similarly to above, `y[-1]`?

File: pycc/rt/rtcc.py [hunk 4]
```diff
@@ -292,6 +275,67 @@ def lagrangian(self, t, t1, t2, l1, l2):
...
+
+        B = 1
+        B -= contract("ia,ia->", l1_r, (t1_r - t1_l))
+        B -= 0.5*contract("ijab,ijab->", l2_r, (t2_r - t2_l))
+        B += 0.5*contract("ijab,ia,jb->", l2_r, t1_r, t1_r)
+        B += 0.5*contract("ijab,ia,jb->", l2_r, t1_l, t1_l)
+        B -= contract("ijab,ia,jb->", l2_r, t1_l, t1_r)
+        B *= np.exp(-phase_r) * np.exp(phase_l)
+
+        return 0.5*A + 0.5*np.conj(B)
```
Comment (bgpeyton) [inline pycc/rt/rtcc.py:337]: is this computing a symmetrized version of Eq. 19 from [J. Chem. Phys. 150, 144106 (2019)](https://doi.org/10.1063/1.5085390)?

**Review by bgpeyton — CHANGES_REQUESTED** [review]

File: pycc/rt/rtcc.py (hunk 4 above)
Comment (lothian) [reply pycc/rt/rtcc.py:337]: No, it's actually Eq. (18).  Eq. (19) is just used in the paper to make an argument about the connection between the autocorrelation function and the ground-state population.

**Review by bgpeyton — APPROVED** [review]
LGTM!

**Review by zw0930 — APPROVED** [review]

**Pull request merged**